### PR TITLE
Allow users to get error locations

### PR DESCRIPTION
**Pull request by cetra3**
Simple change to make ErrorImpl public.

I'd like to be able to use the Marker Information from the scanner error to provide users that are changing files the ability to check to see whether their yaml is valid and, if not, where the issue lies.  To do this i need the Marker information from the ScannerError in a structured format, so I can feed that into an editor.

This resolves https://github.com/dtolnay/serde-yaml/issues/78

I also have a [PR out](https://github.com/chyh1990/yaml-rust/pull/90) about getting access to the Marker for Scanner errors, which is also not public.

### CODE REVIEW

**Comment by cetra3** [comment]
Ok, I've added two new structs to the Error which decouple the Error from yaml_rust in the future.  They are `ErrorMarker` which is just `Marker`, and `ErrorPos` which is a description with an optional ErrorMarker. 

The ErrorPos struct is basically what I'm after, so I can get a nice location of where the error is happening (if one exists).

Unfortunately though, yaml_rust does not currently expose a way of getting the marker from a scan error, which I have submitted a PR for, but may not get accepted, so I've just pointed this PR to my branch.

But to make matters worse, the new version of yaml_rust escapes `y` as a key with quotes around (i.e, `"y"` instead of `y`) I think to align to the [bool](http://yaml.org/type/bool.html) definition within yaml which breaks assertions and examples.

**Comment by dtolnay** [comment]
Please rebase on the master branch which includes the update to yaml-rust 0.4.0.

File: src/error.rs [hunk 1]
```diff
@@ -19,11 +19,63 @@ use yaml_rust::scanner::{self, Marker, ScanError};
...
+    pub fn line(&self) -> usize {
+        self.line
+    }
+
+    pub fn col(&self) -> usize {
+        self.col
+    }
+}
+
+impl<'a> From<&'a Marker> for ErrorMarker {
```
Comment (dtolnay) [inline src/error.rs]: This is still yaml-rust in the public API of serde-yaml. Imagine this code:

```toml
[dependencies]
yaml-rust = "0.3"
serde_yaml = "0.7"
```

```rust
use yaml_rust::scanner::Marker;
use serde_yaml::ErrorMarker;

fn x(marker: &Marker) -> ErrorMarker {
    From::from(marker)
}
```

This code would fail to compile after we bump yaml-rust version.

**Comment by cetra3** [comment]
I've adjusted the from to be a private fn of ErrorMarker rather than implementing the From trait.  Is this enough to hide it?

**Comment by cetra3** [comment]
I've updated it to be a bit cleaner and provide some more documentation around usage.

File: src/error.rs [hunk 2]
```diff
@@ -49,7 +49,72 @@ pub struct Pos {
...
+    pub fn line(&self) -> usize {
+        self.line
+    }
+
+    /// The column of the error
+    pub fn col(&self) -> usize {
+        self.col
+    }
+
+    fn from_marker(marker: &Marker) -> Self {
```
Comment (killercup) [inline src/error.rs]: Can you add a comment that this is required to keep Marker private?

Alternatively, you can impl From and add `#[doc(hidden)]`.

File: src/error.rs [hunk 3]
```diff
@@ -49,7 +49,72 @@ pub struct Pos {
...
+    /// # 
+    /// // The `@` character as the first character makes this invalid yaml
+    /// let result: Result<Value, Error> = yaml(r#"@invalid_yaml"#);
+    /// 
+    /// let marker = result.unwrap_err().marker().unwrap();
+    ///
+    /// assert_eq!(marker.line(), 1);
+    /// assert_eq!(marker.col(), 0);
+    ///
+    ///
```
Comment (killercup) [inline src/error.rs]: Why the empty lines?

File: src/error.rs [hunk 4]
```diff
@@ -49,7 +49,72 @@ pub struct Pos {
...
+            line: marker.line(),
+        }
+    }
+}
+  
 impl Error {
+    /// Returns the ErrorMarker from the error if one exists.
+    /// Allows for the exact location of the error to be gathered.
+    /// Not all errors have a location, so this will return `None`
+    /// # Examples
```
Comment (killercup) [inline src/error.rs]: Please add empty line above headline

File: src/error.rs [hunk 5]
```diff
@@ -49,7 +49,72 @@ pub struct Pos {
...
+            col: marker.col(),
+            index: marker.index(),
+            line: marker.line(),
+        }
+    }
+}
+  
 impl Error {
+    /// Returns the ErrorMarker from the error if one exists.
+    /// Allows for the exact location of the error to be gathered.
```
Comment (killercup) [inline src/error.rs]: If you add an empty line above this one, the first line becomes a short and sweet summary in the API docs :)

File: src/error.rs [hunk 6]
```diff
@@ -49,7 +49,72 @@ pub struct Pos {
...
+    /// Not all errors have a location, so this will return `None`
```
Comment (killercup) [inline src/error.rs]: Please end sentence with a period

File: src/error.rs [hunk 7]
```diff
@@ -49,7 +49,72 @@ pub struct Pos {
...
+    /// # Examples
+    ///
+    /// ```rust
+    /// # extern crate serde_yaml;
+    /// # use serde_yaml::{Value, Error};
+    ///
```
Comment (killercup) [inline src/error.rs]: As the lines above and below are hidden in the docs, this code block will start with an empty line. Speaking of empty lines, can you remove this one? 😉

File: src/error.rs [hunk 8]
```diff
@@ -49,7 +49,72 @@ pub struct Pos {
...
+    /// # fn yaml(i: &str) -> Result<Value, Error> { serde_yaml::from_str(i) }
+    /// # fn main() {
+    /// # 
+    /// // The `@` character as the first character makes this invalid yaml
+    /// let result: Result<Value, Error> = yaml(r#"@invalid_yaml"#);
```
Comment (killercup) [inline src/error.rs]: No need to use a `r#"raw"#` string here, right?

File: src/error.rs [hunk 9]
```diff
@@ -49,7 +49,72 @@ pub struct Pos {
...
+    /// Returns the ErrorMarker from the error if one exists.
+    /// Allows for the exact location of the error to be gathered.
+    /// Not all errors have a location, so this will return `None`
+    /// # Examples
+    ///
+    /// ```rust
+    /// # extern crate serde_yaml;
+    /// # use serde_yaml::{Value, Error};
+    ///
+    /// # fn yaml(i: &str) -> Result<Value, Error> { serde_yaml::from_str(i) }
```
Comment (killercup) [inline src/error.rs]: I'm not sure this is necessary, as `yaml` is only ever used once in this example. And `serde_yaml::from_str` is not that verbose, is it?

File: src/error.rs [hunk 10]
```diff
@@ -49,7 +49,72 @@ pub struct Pos {
...
+    /// assert_eq!(marker.col(), 0);
+    ///
+    ///
+    /// # }
+    /// ```
+    pub fn marker(&self) -> Option<ErrorMarker> {
+        match *self.0 {
+            ErrorImpl::Message(_, Some(ref pos)) => Some(ErrorMarker::from_marker(&pos.marker)),
+            ErrorImpl::Scan(ref scan) => Some(ErrorMarker::from_marker(scan.marker())),
+            _ => None
```
Comment (killercup) [inline src/error.rs]: pls add trailing comma

File: src/error.rs [hunk 11]
```diff
@@ -49,7 +49,72 @@ pub struct Pos {
...
+    /// # fn yaml(i: &str) -> Result<Value, Error> { serde_yaml::from_str(i) }
+    /// # fn main() {
+    /// # 
+    /// // The `@` character as the first character makes this invalid yaml
+    /// let result: Result<Value, Error> = yaml(r#"@invalid_yaml"#);
+    /// 
+    /// let marker = result.unwrap_err().marker().unwrap();
+    ///
+    /// assert_eq!(marker.line(), 1);
+    /// assert_eq!(marker.col(), 0);
```
Comment (killercup) [inline src/error.rs]: It's weird to see lines start at 1 but cols at zero… but whatever

**Comment by dtolnay** [comment]
How come the column does not match what you see when you print the error?

```rust
extern crate serde_yaml;
use serde_yaml::Value;

fn main() {
    let err = serde_yaml::from_str::<Value>("@").unwrap_err();
    let marker = err.marker().unwrap();
    println!("{}", err);
    println!("line={} col={}", marker.line(), marker.col());
}
```

```
unexpected character: `@' at line 1 column 1
line=1 col=0
```

File: src/error.rs [hunk 12]
```diff
@@ -49,7 +49,72 @@ pub struct Pos {
     path: String,
 }
 
+
+/// This type represents the location that an error occured
+#[derive(Debug)]
+pub struct ErrorMarker {
```
Comment (dtolnay) [inline src/error.rs]: Could you brainstorm some alternative names for this? [`marker`](https://doc.rust-lang.org/std/marker/index.html) means something else in Rust.

File: src/error.rs [hunk 13]
```diff
@@ -49,7 +49,72 @@ pub struct Pos {
...
+        self.index
+    }
+
+    /// The line of the error
+    pub fn line(&self) -> usize {
+        self.line
+    }
+
+    /// The column of the error
+    pub fn col(&self) -> usize {
```
Comment (dtolnay) [inline src/error.rs]: I would prefer not to abbreviate "column" because the standard library does not abbreviate it [here](https://doc.rust-lang.org/std/macro.column.html) and [here](https://doc.rust-lang.org/std/panic/struct.Location.html#method.column).

**Comment by cetra3** [comment]
@dtolnay I'm not sure about why the marker is different from the error. I'll have a look at the yaml crate and see if I can pin that one down.

@killercup I'll make those adjustments and check that it formats it as you said.

**Comment by cetra3** [comment]
I've pushed a new commit to the branch with some of those suggestions.

I've changed the name simply to `Location` and used `column()` rather than `col`.

With regards to the column not matching the error message, I've just added + 1 to what we get from the marker, as that is what they're doing when they print it out in the [yaml crate](https://github.com/chyh1990/yaml-rust/blob/master/src/scanner.rs#L83).

Also added a test case for the None side of things.

**Comment by dtolnay** [comment]
I will get a release out tomorrow with this change and the update to yaml-rust 0.4.

**Comment by cetra3** [comment]
Awesome!
